File: buildSrc/src/main/java/org/stepik/gradle/plugins/jetbrains/PatchPluginXmlTask.java
package org.stepik.gradle.plugins.jetbrains;

import com.intellij.structure.domain.IdeVersion;
import org.gradle.api.file.FileCollection;
import org.gradle.api.internal.ConventionTask;
import org.gradle.api.tasks.Input;
import org.gradle.api.tasks.InputFiles;
import org.gradle.api.tasks.Optional;
import org.gradle.api.tasks.OutputDirectory;
import org.gradle.api.tasks.SkipWhenEmpty;
import org.gradle.api.tasks.TaskAction;
import org.jdom2.Document;
import org.jdom2.Element;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;

/**
 * @author meanmail
 */
public class PatchPluginXmlTask extends ConventionTask {
    private String destinationDirName;
    private ProductPluginExtension extension;

    @SuppressWarnings("WeakerAccess")
    @OutputDirectory
    public File getDestinationDir() {
        return destinationDirName != null ? new File(getProject().getBuildDir(), destinationDirName) : null;
    }

    @SuppressWarnings("WeakerAccess")
    @SkipWhenEmpty
    @InputFiles
    public FileCollection getPluginXmlFiles() {
        return Utils.sourcePluginXmlFiles(getProject());
    }

    @SuppressWarnings("WeakerAccess")
    @Input
    @Optional
    public String getVersion() {
        return getProject().getVersion().toString();
    }

    @SuppressWarnings("WeakerAccess")
    @Input
    @Optional
    public String getPluginDescription() {
        return extension != null ? extension.getPluginDescription() : null;
    }

    @SuppressWarnings("WeakerAccess")
    @Input
    @Optional
    public String getSinceBuild() {
        if (extension.getUpdateSinceUntilBuild()) {
            IdeVersion ideVersion = IdeVersion.createIdeVersion(extension.getDependency().getBuildNumber());
            return ideVersion.getBaselineVersion() + "." + ideVersion.getBuild();
        }

        return null;
    }

    @SuppressWarnings("WeakerAccess")
    @Input
    @Optional
    public String getUntilBuild() {
        if (extension.getUpdateSinceUntilBuild()) {
            IdeVersion ideVersion = IdeVersion.createIdeVersion(extension.getDependency().getBuildNumber());

            if (extension.getSameSinceUntilBuild()) {
                return ideVersion.getBaselineVersion() + ".*";
            } else if (extension.getUntilBuild() != null) {
                return extension.getUntilBuild() + ".*";
            }
        }
        return null;
    }

    @SuppressWarnings("WeakerAccess")
    @Input
    @Optional
    public String getChangeNotes() {
        return extension != null ? extension.getChangeNotes() : null;
    }

    @TaskAction
    public void patchPluginXmlFiles() {
        FileCollection files = getPluginXmlFiles();
        if (files == null) {
            return;
        }

        files.forEach(file -> {
            try {
                Document pluginXml = Utils.getXmlDocument(file);
                if (pluginXml == null) {
                    return;
                }
                patchSinceUntilBuild(pluginXml, getSinceBuild(), getUntilBuild());
                patchElement(pluginXml, "description", getPluginDescription());
                patchElement(pluginXml, "change-notes", getChangeNotes());
                patchElement(pluginXml, "version", getVersion());

                File destinationFile = new File(getDestinationDir(), file.getName());

                Utils.outputXml(pluginXml, destinationFile);
            } catch (IOException ignore) {
            }
        });
    }

    private void patchSinceUntilBuild(
            @NotNull Document pluginXml,
            @Nullable String sinceBuild,
            @Nullable String untilBuild) {
        if (extension.isUpdateSinceUntilBuild()) {
            Element result = pluginXml.getRootElement().getChild("idea-version");

            if (result != null) {
                Utils.setAttributeValue(result, "since-build", sinceBuild);
                Utils.setAttributeValue(result, "until-build", untilBuild);
            }
        }
    }

    private static void patchElement(@NotNull Document pluginXml, @NotNull String name, @Nullable String value) {
        if (value != null) {
            Element result = pluginXml.getRootElement().getChild(name);

            if (result == null) {
                result = new Element(name);
                pluginXml.getRootElement().addContent(result);
            }

            result.setText(value);
        }
    }

    void setDestinationDirName(String destinationDirName) {
        this.destinationDirName = destinationDirName;
    }

    void setExtension(ProductPluginExtension extension) {
        this.extension = extension;
    }
}

File: buildSrc/src/main/kotlin/org/stepik/gradle/plugins/jetbrains/tasks/PatchPluginXmlTask.kt
package org.stepik.gradle.plugins.jetbrains.tasks

import org.gradle.api.file.FileCollection
import org.gradle.api.internal.ConventionTask
import org.gradle.api.tasks.Input
import org.gradle.api.tasks.InputFiles
import org.gradle.api.tasks.Optional
import org.gradle.api.tasks.OutputDirectory
import org.gradle.api.tasks.SkipWhenEmpty
import org.gradle.api.tasks.TaskAction
import org.jdom2.CDATA
import org.jdom2.Content
import org.jdom2.Document
import org.jdom2.Element
import org.jdom2.Text
import org.stepik.gradle.plugins.jetbrains.ProductPluginExtension
import org.stepik.gradle.plugins.jetbrains.Utils
import org.stepik.gradle.plugins.jetbrains.Utils.sourcePluginXmlFiles
import java.io.File
import java.io.IOException


open class PatchPluginXmlTask : ConventionTask() {

    var destinationDirName: String? = null

    var extension: ProductPluginExtension? = null

    @get:[Input Optional]
    val version: String
        get() = project.version.toString()

    @get:OutputDirectory
    private val destinationDir: File?
        get() {
            destinationDirName ?: return null
            return File(project.buildDir, destinationDirName)
        }

    @get:[SkipWhenEmpty InputFiles]
    private val pluginXmlFiles: FileCollection
        get() {
            return sourcePluginXmlFiles(project)
        }

    @get:[Input Optional]
    private val pluginDescription: String?
        get() = extension?.pluginDescription

    @get:[Input Optional]
    private val sinceBuild: String?
        get() = extension?.sinceBuild

    @get:[Input Optional]
    private val untilBuild: String?
        get() = extension?.untilBuild

    @get:[Input Optional]
    private val changeNotes: String?
        get() = extension?.changeNotes

    @TaskAction
    fun patchPluginXmlFiles() {
        pluginXmlFiles.forEach {
            try {
                val pluginXml = Utils.getXmlDocument(it) ?: return

                patchSinceUntilBuild(pluginXml, sinceBuild, untilBuild)
                patchElement(pluginXml, "description", CDATA(pluginDescription))
                patchElement(pluginXml, "change-notes", CDATA(changeNotes))
                patchElement(pluginXml, "version", version)

                val destinationFile = File(destinationDir, it.name)

                Utils.outputXml(pluginXml, destinationFile)
            } catch (ignore: IOException) {
            }
        }
    }

    private fun patchSinceUntilBuild(pluginXml: Document, sinceBuild: String?, untilBuild: String?) {
        if (extension?.updateSinceUntilBuild != true) {
            return
        }

        val result = pluginXml.rootElement.getChild("idea-version")

        if (result != null) {
            Utils.setAttributeValue(result, "since-build", sinceBuild)
            Utils.setAttributeValue(result, "until-build", untilBuild)
        }
    }

    companion object {

        fun patchElement(pluginXml: Document, name: String, value: String?) {
            patchElement(pluginXml, name, Text(value))
        }

        fun patchElement(pluginXml: Document, name: String, value: Content?) {
            if (value != null) {
                var result = pluginXml.rootElement.getChild(name)

                if (result == null) {
                    result = Element(name)
                    pluginXml.rootElement.addContent(result)
                }

                result.setContent(value)
            }
        }
    }
}
